### PR TITLE
WIP: Templating: Moves QueryVariable to Redux

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Third iteration trying to move template variables to Redux. This iteration tries to replace QueryVariable vertically through all layers and add an adapter pattern that can be used to move other variable types moving forward. 

By vertically I mean that we're trying to integrate QueryVariable type into state and use React to render from state. In this #20155 PR I tried an horizontal approach that replaces the whole state layer for all variable types and leaving Angular to render from state.

This PR introduces an adapter pattern with this interface (**subject to change before closing this PR**)
```typescript
export interface VariableAdapter<Model extends VariableModel, State extends VariableState> {
  description: string;
  dependsOn: (variable: Model, variableToTest: Model) => boolean;
  setValue: (variable: Model, option: VariableOption) => Promise<void>;
  setValueFromUrl: (variable: Model, urlValue: UrlQueryValue) => Promise<void>;
  updateOptions: (variable: Model, searchFilter?: string) => Promise<void>;
  onEditorUpdate: (variable: Model) => Promise<void>;
  picker: ComponentType<VariableState>;
  editor: ComponentType<VariableEditorProps>;
  reducer: Reducer<State, PayloadAction<VariablePayload<any>>>;
}
```

The adapter forces each variable type to implement its own `reducer`, `picker` and `editor` to separate logic, state and rendering.

VariablesInRedux

The goal is prove that the adapter pattern is a valid approach but there's still a lot of work to do:

**Core**
- [x] fix more performant state
- [x] remove variable lookup loop from templateSrv
- [x] fix duplicate variable
- [x] fix delete variable
- [x] fix reorder variables
- [x] add new variable
- [x] fix when dashboard only contains adapted variables it doesn't show submenu
- [x] fix save dashboard
- [x] fix so we cleanup state when dashboard is unloaded  
- [x] add getValueFromUrl flow from old variables
- [x] validate that url changes work
- [ ] fix dependencies between angular variable and redux variable and vice versa 

**Picker**
- [x] fix tags ui in Picker
- [x] fix keydown handling in Picker.
- [x] fix clear selections in Picker
- [x] fix searchFilter lazy load of options in Picker image image
- [x] fix small bug with multiple dropdowns can be opened at the same time in Picker 

**Misc**
- [x] fix tests
- [x] fix strict null errrors
- [ ] write tests
- [ ] cleanup/refactor code
- [ ] update e2e tests
- [ ] documentation
- [ ] celebrate 🎉 

**Known issues so far:**
- [ ] Some kind of infinite loop when changing data source type
- [ ] If I change the selected value of the server variable I can see redux actions for updating the options for the datacenter value being fired as well.
- [ ] focus is not following highlighted picker option when navigating with keys (same on play.grafana.com)

**When we have nothing else to do:**
- [x] Simplify state handling
- [ ] Create a tag component to handle the picker tags
- [ ] Change so the VariableTag and VariableOption uses text/value: string[] instead of string | string[]
- [ ] Change so thunk signatures take `uuid` as parameter instead of `variable` to avoid stupid bugs
- [ ] Write a class that walks the variable tree and create an execution order.

**Upcoming PRs**
- [ ] Refactor Picker/Picker actions and state into something that can be reused between custom/query/datasource variable types
- [x] Can we simplify state handling and use only the variable type specific reducer?
- [ ] Introduce custom variable type
- [ ] Introduce data source variable type
- [ ] Introduce ad hoc variable type
- [ ] Introduce interval variable type
- [ ] Introduce custom variable type
- [ ] Introduce constant variable type

**Which issue(s) this PR fixes**:
Related #19896

**Special notes for your reviewer**:
Not covered in this PR
- #20703
- New Form Styles

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
Wow, so much here :) 

Some great progress on migrating the variable select picker to react as well! 

Looking very promising. Wonder how we can reuse reducer & picker state & component between types, as they will be close to identical for some variables (say custom & with multi select). So actions like showQueryVariableDropDown or hideQueryVariableDropDown maybe is a bit too specific to a single variable type? The editor will be shared for custom, data source, query,  


Not sure how ready for testing it is but found some issues.

When on http://localhost:3000/d/-Y-tnEDWk/templating-nested-template-variables 
1) Creating new variable, the view is empty
2) Some kind of infinite loop when changing data source type 
3) If I change the selected value of the server variable I can see redux actions for updating the options for the datacenter value being fired as well.

File: public/app/features/templating/state/queryVariableReducer.ts [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+    const results = action.payload.data;
+    const tags: VariableTag[] = [];
+    for (let i = 0; i < results.length; i++) {
+      tags.push({ text: results[i].text, selected: false });
+    }
+
+    return { ...state, variable: { ...state.variable, tags } };
+  }
+
+  if (setCurrentVariableValue.match(action)) {
```
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/features/templating/state/queryVariableReducer.ts]: I see, I haven't put so much thought into that yet. I guess we can reuse actions/reducers. I wanted to have a clear separation between the types so they don't get mixed up like today but any ideas on how to separate these are welcome.

File: public/app/features/templating/editor/QueryVariableEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+                    {''}
+                  </option>
+                  {this.props.dataSources.length &&
+                    this.props.dataSources.map(ds => (
+                      <option key={ds.value} value={ds.value} label={ds.name}>
+                        {ds.name}
+                      </option>
+                    ))}
+                </select>
+              </div>
```
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/features/templating/editor/QueryVariableEditor.tsx]: Yes, this field need to be requires and DSPicker didn't have support for it so I didn't want to expand the PR further. But you're right all this should be replaced with new Form styles.

File: public/app/features/templating/state/selectors.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { StoreState } from '../../../types';
+import { VariableModel } from '../variable';
+import { getState } from '../../../store/store';
+
+export const getVariableByName = <T extends VariableModel = VariableModel>(
+  name: string,
+  state: StoreState = getState()
+): T => {
+  const variable = getVariables(state).find(v => v.name === name);
```
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/features/templating/state/selectors.ts]: Good catch, we need to store a lookup structure with both `name` and `id` to makes this lightning fast.

**Comment by hugohaggmark** [comment]
> Wow, so much here :)
> 
> Some great progress on migrating the variable select picker to react as well!
> 
> Looking very promising. Wonder how we can reuse reducer & picker state & component between types, as they will be close to identical for some variables (say custom & with multi select). So actions like showQueryVariableDropDown or hideQueryVariableDropDown maybe is a bit too specific to a single variable type? The editor will be shared for custom, data source, query,
> 
> Not sure how ready for testing it is but found some issues.
> 
> When on http://localhost:3000/d/-Y-tnEDWk/templating-nested-template-variables
> 
> 1. Creating new variable, the view is empty

Yeah, the Add is not supported yet. Adding that to the checklist.
> 2. Some kind of infinite loop when changing data source type

I couldn't repro this but will add it to the checklist as well.
> 3. If I change the selected value of the server variable I can see redux actions for updating the options for the datacenter value being fired as well.

I couldn't repro this neither and it's using the Graph as previously but I'll add it to the checklist

**Comment by torkelo** [comment]
Could be useful to add the custom variable type to this as well to see how more reuse can be had, as the picker and selection actions are identical between many of the variables

**Comment by torkelo** [comment]
But that could be added & refactored in follow-up PRs as well. Think the biggest problem I see with this is that too much is query variable specific and there going to be a too high amount of duplication between variable types (that mostly only differ in how they get options)

File: public/app/features/templating/state/index.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            isValid: false,
+            errors: {
+              ...state.variables[action.payload.uuid].editor.errors,
+              update: action.payload.data.message,
+            },
+          },
+        },
+      },
+    };
+  }
```
Comment (torkelo) [inline public/app/features/templating/state/index.ts]: hope we can use immer reducers later on when all state mutation is removed :)

**Comment by hugohaggmark** [comment]
> But that could be added & refactored in follow-up PRs as well. Think the biggest problem I see with this is that too much is query variable specific and there going to be a too high amount of duplication between variable types (that mostly only differ in how they get options)

Yes, I think introducing `custom` type as well is a good idea to iron out any duplications.

**Pull request closed without merging**